Add dateinterval type reference

MisatoTremor opened the pull request:
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | 3.2 |
| Fixed tickets | #4816 |

MisatoTremor committed:
Add dateinterval type reference

xabbuh commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,312 @@
+| with_seconds | ``Boolean`` | The value of the `with_seconds`_ option.                        |
++--------------+-------------+-----------------------------------------------------------------+
+| with_invert  | ``Boolean`` | The value of the `with_invert`_ option.                         |
++--------------+-------------+-----------------------------------------------------------------+
```
I would use the "simple" table syntax:

```rst
==============  ===========  ======================================
Variable        Type         Usage
==============  ===========  ======================================
``widget``      ``mixed``    The value of the `widget`_ option.
``with_years``  ``boolean``  The value of the `with_years`_ option.
and             so           on
==============  ===========  ======================================
```

MisatoTremor replied:
Done

xabbuh commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,312 @@
+
+**type**: ``boolean`` **default**: ``false``
+
+Default value of the invert field type::
```
What does this option actually mean?

MisatoTremor replied:
Was an obsolete option, now removed

javiereguiluz commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,294 @@
+The format of the *input* data - i.e. the format that the interval is stored on
+your underlying object. Valid values are:
+
+* ``string`` (e.g. ``P1DT12H``)
```
Reading the introduction of this article, it seems that the string must be formatted using `RFC 3339`. If this is correct, I think we should change this phrase by the following:

```
* ``string`` (a string formatted with `RFC 3339`_ standard, e.g. ``P1DT12H``)
```

javiereguiluz commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,294 @@
+=======================
+
+This field type allows the user to modify data that represents a specific
+interval (e.g. ``P7Y6M5DT12H15M30S``).
```
Displaying this strange interval string at the beginning of the article seems a bit odd to me. What about rewording the first two sentences as follows?

```
This field type allows the user to modify data that represents a specific
date interval. It can be rendered as a integer or text input or select tags.

The underlying format of the data can be a ``DateInterval`` object,
a `RFC 3339`_ duration string (e.g. ``P7Y6M5DT12H15M30S``) or an array.
```

MisatoTremor commented:
Changed it like @javiereguiluz said.
I only swapped the two RFC duration strings, putting the shorter (and hopefully easier to understand) one in the intro text and the full example in the string input description.
Also fixed the sorting of the options list to match the other type docs.

HeahDude commented:
https://github.com/symfony/symfony/pull/16809#issuecomment-227728131 is merged now, can a maintainer please remove "On hold" label?

@MisatoTremor What the status here?

MisatoTremor commented:
@HeahDude 
Updated for 3.2

xabbuh commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,293 @@
+| Options              | - `days`_                                                                         |
+|                      | - `placeholder`_                                                                  |
+|                      | - `hours`_                                                                        |
+|                      | - `input`_                                                                        |
```
looks like something went wrong here

MisatoTremor replied:
Removed read_only field reference, build now runs successfully.

@xabbuh What do you mean with this line comment?

xabbuh replied:
@MisatoTremor When I reviewed this earlier today the diff looked completely broken. Looks like this was related to my browser then or the issue disappeared with your push. So everything's fine here how. :)

xabbuh commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,293 @@
+
+.. include:: /reference/forms/types/options/mapped.rst.inc
+
+.. include:: /reference/forms/types/options/read_only.rst.inc
```
The `read_only` option does not exist anymore.

WouterJ commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,290 @@
+
+DateIntervalType Field Type
+===========================
+
```
missing a versionadded directive:

```rst
.. versionadded:: 3.2
   The DateIntervalType field type was introduced in Symfony 3.2.
```

MisatoTremor replied:
Added

WouterJ commented:
👍 Thanks for writing the documentation for the feature @MisatoTremor!

status: reviewed

javiereguiluz commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,290 @@
+date interval. It can be rendered as a integer or text input or select tags.
+
+The underlying format of the data can be a ``DateInterval`` object,
+a `RFC 3339`_ duration string (e.g. ``P1DT12H``) or an array.
```
Are we sure we want to link to the RFC 3339? In that RFC you can read the following:

![rfc_3339](https://cloud.githubusercontent.com/assets/73419/16553141/ba340846-41c6-11e6-8876-0af1e8c02b6c.png)

It looks like we should link instead to ISO 8601 standard: https://en.wikipedia.org/wiki/ISO_8601

javiereguiluz commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,290 @@
+
+**type**: ``string`` **default**: ``choice``
+
+Defines the ``widget`` option for the single interval component inputs.
```
I miss the list of possible values for this option. Is just `choice` and `text`?

MisatoTremor replied:
Nope, 2 more. Pushing updated doc in a minute.

javiereguiluz commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,290 @@
+**type**: ``Boolean`` **default**: ``false``
+
+Whether or not to include invert in the input. This will result in an additional
+input to capture seconds.
```
I don't understand this description. What does "invert" mean for the input? And the last sentence seems to be a copy+paste error.

javiereguiluz commented:
@MisatoTremor thanks for adding this form type and for providing the docs too. Today we've highlighted it in the Symfony blog (see [New in Symfony 3.2: DateInterval form type](http://symfony.com/blog/new-in-symfony-3-2-dateinterval-form-type)).

When writing the blog post, I wondered if this type could be a bit easier to use. In particular, I wonder if the `input` option could be removed. If it's only used to tell the form type which format we are using (DateInterval, array or string) I think we can detect the format automatically. Do you think that would be possible? Thanks!

MisatoTremor commented:
@javiereguiluz Thanks for the nice highlight! :)

I've included the option according to other date/time types, but this indeed sounds like a great improvement. Maybe this could be done for the other types, too. Will look into into

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+options are ``input`` and ``widget``.
+
+Suppose that you have a ``remindEvery`` field whose underlying interval is a
+``DateInterval`` object. The following configures the ``dateinterval`` type
```
```rst
... configures the ``DateIntervalType``
```

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+.. code-block:: php
+
+    $builder->add('remindEvery', DateIntervalType::class, array(
+        'widget' => 'choice',
```
Should be `ChoiceType::class`, correct? And we usually add the `use` statements above :)

```rst
 use Symfony\Component\Form\Extension\Core\Type\DateIntervalType;
 // ...

$builder...
```

weaverryan replied:
oh no, it shouldn't be `ChoiceType::class` - I see now :)

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+.. index::
+   single: Forms; Fields; DateIntervalType
+
+DateIntervalType Field Type
```
We don't usually include the last word `Type` - so just `DateIntervalType Field`

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+   The DateIntervalType field type was introduced in Symfony 3.2.
+
+This field type allows the user to modify data that represents a specific
+date interval. It can be rendered as a integer or text input or select tags.
```
Since this is repeated below, how about just:

> It can be rendered in a variety of different ways.

weaverryan replied:
Actually, let's re-word everything. I realized I wasn't fully understanding *when* this field would
be useful at first (i was thinking "date range" at first, not duration).

```rst
This field allows the user to select an *interval* of time. For example, if you want to
allow the user to choose *how often* they receive a status email, they could use this
field to choose intervals like every "10 minutes" or "3 days".

The field can be rendered in a variety of different ways (see `widget`_) and can be configured to
give you a ``DateInterval`` object, an `ISO 8601`_ duration string (e.g. ``P1DT12H``)
or an array (see `input`_).
```

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
++----------------------+-----------------------------------------------------------------------------------+
+| Underlying Data Type | can be ``DateInterval``, string or array (see the ``input`` option)               |
++----------------------+-----------------------------------------------------------------------------------+
+| Rendered as          | single text box or up to six select, text or integer boxes plus optional checkbox |
```
single text box, multiple text boxes or select fields - see the `widget`_ option

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+-----------
+
+This field type is highly configurable, but easy to use. The most important
+options are ``input`` and ``widget``.
```
Let's make `input` and `widget` links

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+
+Suppose that you have a ``remindEvery`` field whose underlying interval is a
+``DateInterval`` object. The following configures the ``dateinterval`` type
+for that field as **three different choice fields**:
```
Let's clarify *what* fields are shown by default:

```rst
**three different choice fields**: year, month and day:
```

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+        'widget' => 'choice',
+    ));
+
+The field also supports an ``array`` as valid ``input`` option values.
```
How about we replace this whole section with a bigger, example:

```rst
You can configure *a lot* of different options, including exactly *which* range
options to show (e.g. don't show "months", but *do* show "days")::

$builder->add('remindEvery', DateIntervalType::class, array(
    'widget'      => 'integer', // render a text field for each part
    // 'input'    => 'string',  // if you want the field to return a ISO 8601 string back to you

    // customize which text boxes are shown
    'with_years'  => false,
    'with_months' => false,
    'with_days'   => true,
    'with_hours'  => true,
));
```

The `with_*` stuff seems a bit weird to use :). Did you consider adding something like a `components` option instead? e.g.

```php
'components' => ['days', 'hours', 'minutes']
```

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+|                      | - `with_seconds`_                                                                 |
+|                      | - `with_weeks`_                                                                   |
+|                      | - `with_years`_                                                                   |
+|                      | - `years`_                                                                        |
```
Are we typically making these options alphabetical now? I don't think we should - I think we should order them by importance. So, something more like:

* widget
* input
* with_*
* placeholder
* then all of the customization for the different values - e.g. `years`, etc

MisatoTremor replied:
Back then I used the datetime doc as base for my documentation which had it that way. Some types are alphabetically, others aren't.
Maybe that's something to rework for all types in another PR?

weaverryan commented on reference/forms/types/dateinterval.rst:
```diff
@@ -0,0 +1,325 @@
+|                      | - `invalid_message_parameters`_                                                   |
+|                      | - `mapped`_                                                                       |
++----------------------+-----------------------------------------------------------------------------------+
+| Parent type          | :doc:`form </reference/forms/types/form>`                                         |
```
FormType

weaverryan commented:
Thanks for the work @MisatoTremor - it's a really interesting component! And I'm sorry, I just peppered you with comments :)

Status: Needs Work

MisatoTremor commented:
@weaverryan Thank for your review Ryan. I applied all but one which I commented upon.

wouterj commented:
Hi @MisatoTremor! I'm sorry for the long delay, seems like this one slipped under our radar. As 3 doc members already reviewed this and you fixed their comments, I assume this one is ready to merge.

/cc @symfony/team-symfony-docs 

status: reviewed

javiereguiluz commented:
👍  this is indeed ready to be merged. Thanks @MisatoTremor.

MisatoTremor commented:
@wouterj I totally understand this. :) There's much to do and also more important sections. You all are doing a great job!

Thanks for the update.

wouterj commented:
Thanks Steffen for your work & patience! This is now merged in the docs and will be online tomorrow.
